go-controller/pkg/crd/egressfirewall/v1/types.go

```diff
@@ -0,0 +1,78 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EgressNetworkFirewallRuleType indicates whether an EgressNetworkFirewallRule allows or denies traffic
+type EgressFirewallRuleType string
+
+const (
+	EgressFirewallRuleAllow EgressFirewallRuleType = "Allow"
+	EgressFirewallRuleDeny  EgressFirewallRuleType = "Deny"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +resource:path=egressfirewall
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// EgressFirewall describes the current egress firewall for a Namespace.
+// Traffic from a pod to an IP address outside the cluster will be checked against
+// each EgressFirewallRule in the pod's namespace's EgressFirewall, in
+// order. If no rule matches (or no EgressFirewall is present) then the traffic
+// will be allowed by default.
+type EgressFirewall struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Specification of the desired behavior of EgressFirewall.
+	Spec EgressFirewallSpec `json:"spec"`
+	// Observed status of EgressFirewall
+	// +optional
+	Status EgressFirewallStatus `json:"status"`
+}
+
+// EgressFirewallStatus describes the lifecycle status of EgressFirewall.
+type EgressFirewallStatus struct {
+}
+
+// EgressFirewallSpec is a desired state description of EgressFirewall.
+type EgressFirewallSpec struct {
+	// a collection of egress firewall rule objects
+	Rules []EgressFirewallRule `json:"rules"`
+}
+
+// EgressFirewallRule is a single egressfirewall rule object
+type EgressFirewallRule struct {
+	// type marks this as an "Allow" or "Deny" rule
+	Type EgressFirewallRuleType `json:"type"`
+	// protocol marks what protocol for the rule to affect
+	EgressFirewallProtocol `json:"protocoldetails,omitempty"`
+	// to is one or more targes that traffic is allowed/denied to
+	To []Destination `json:"to"`
+}
+
+//EgressFirewallProtocol specifies the protocol and optionally port number to allow or deny traffic
+type EgressFirewallProtocol struct {
+	Protocol string `json:"protocol"`
+	Port     int32  `json:"port,omitempty"`
```

Comment:
so again, `EgressFirewallPort` since it's mostly about the port, not the protocol. Add a space after the `//` in the comment.

`Port` should be required and `Protocol` should be optional, and it should default to "TCP" if not specified. Oh, and it should probably be a `v1.Protocol` not a string. (`v1` as in `k8s.io/api/core/v1`). Oh, and you need doc comments explaining the fields.

So basically, this should be almost identical to `NetworkPolicyPort`, except that `Port` should be required and an `int32`. So:

    // EgressFirewallPort specifies the port number to allow or deny traffic to
    type EgressFirewallPort struct {
            // protocol (TCP, UDP, or SCTP) that the traffic must match. Defaults to TCP
            // +optional
            Protocol *v1.Protocol `json:"protocol,omitempty"`

            // port that the traffic must match
            Port int32  `json:"port"`
    }

(in particular note the `*` in `*v1.Protocol`; to make it properly optional it needs to be a pointer value in the golang struct)

Reply:
changed